tests/conftest.py

```diff
@@ -127,17 +169,66 @@ def cert_encrypted_private_key_file(example_cert):
 
 
 @pytest.fixture
-async def server():
+def restart_queue():
+    return asyncio.Queue()
+
+
+@pytest.fixture
+def restart_server(restart_queue, backend):
+    """Restart the running server from an async test function."""
+
+    async def asyncio_restart():
+        await restart_queue.put(None)
+        await restart_queue.get()
+
+    if isinstance(backend, AsyncioBackend):
+        return asyncio_restart
+
+    async def restart():
+        await backend.run_in_threadpool(AsyncioBackend().run, asyncio_restart)
+
+    return restart
+
+
+async def watch_restarts(server, queue):
+    # We can't let async tests run shutdown()/startup() directly because:
+    # * They may be running in a different async environment (e.g. trio).
+    # * When that is the case, that environment runs in a separate thread.
+    # * But shutdown/startup() needs to run on the loop from which the server was
+    # spawned, i.e. the one we're currently in.
+    # * Even if we launched `AsyncioBackend().run` in the threadpool, that would run
+    # on a *new event loop - in yet another thread!
+    # So we use a queue as a thread-safe communication channel with async tests.
```

Comment:
I'm having trouble understanding this flow, let's see if I got it right.

- In master right now we have a single thread that starts the asyncio event loop.
- Each test is an async function that is then scheduled with a server fixture that schedules the server task and yields it to the test.
- Some tests manually call `startup` and `shutdown`.
- On test teardown we signal Uvicorn to exit and await its task.

In this PR we now have multiple concurrency backends, which means:
- If required (i.e. trio) we spawn the environment in a new thread.
- But server fixture cannot be spawned on that separate thread, since it's asyncio only, thus the tests can no longer call `startup` and `shutdown`.
- So to coordinate between test wanting to do this we add the `restart_server` fixture.
- Tests are running now running in the trio thread and when invoking `restart_server` it places `None` in the queue, and blocks until it gets a value.
- This signal is then picked up by `watch_restarts` to now call `server.shutdown` and `server.startup` in the asyncio thread, and finally placing `None` in the queue.
- This is then picked up by `restart_server` and returns.

Is this correct?

Reply:
You’ve got it absolutely right @yeraydiazdiaz! Just a couple of precisions:

> But server fixture cannot be spawned on that separate thread, since it's asyncio only, thus the tests can no longer call startup and shutdown.

The server fixture is always spawned in the main thread, ie in asyncio — and that cannot change AFAIK because of how pytest fixtures work. Rather, the main reason we can no longer call startup/shutdown in async tests is because they may run inside a different async environment — so we can’t just « await » the server’s methods because it server is running in asyncio.

> So to coordinate between test wanting to do this we add the restart_server fixture.

Yes — although resorting to a fixture here is mainly motivated by abstracting all this mess from the test code.

Everything you said about the handling of restart signals is correct. :-)

Your analysis is super helpful, though — I’m thinking we can refine some of the comments based on my answers here so that it’s crystal clear what is happening. Have you got any suggestions on this regard?

Reply:
Also, I have the intuition there may be a better primitively than a queue to implement the « I demand a restart, please notify me when it’s done » flow here.

I think an implementation based on two events (restart_requested and restart_finished) could be more suitable as well as more self-documenting. Basically we would:

- Set restart_requested and wait for restart_finished in restart_server().
- Wait for restart_requested and set restart_finished in watch_restarts().

This implementation should probably be abstracted away from the fixtures as well (eg via a helper class), which would allow us to put all the comments in a single location.